### ISSUE TITLE
Problem with prediction model

### ISSUE THREAD

**Issue by rybickibartek** [aksnzhy/xlearn#204]
Hi,

I built FFM model, CV looks quite ok, loss also. So I decided to make a prediction.
Results are much worse than it is declared in CV - please notice that as a test set I choose train data.

All the predictions have more or less the same value, what looks wired. Target does not look like that.
'-0.386191
-0.385785
-0.385787
-0.385689
-0.385947
-0.385713
-0.386015
-0.385887'

What might be the reason?
Is it some thing related with normalization? In the docs we can find the below sentence:
"Note that if you use Instance-wise Normalization in training process, you also need to use the meachnism in prediction process."

# Code

import xlearn as xl

ffm_model = xl.create_ffm()              
ffm_model.setTrain("./train_ffm.txt")    
ffm_model.setTest("./train_ffm.txt")

param = {'task':'reg', 'lr':0.001, 'lambda':0.0002, 'k':4, 'fold':3}

ffm_model.fit(param, "./model.out")
ffm_model.predict("./model.out", "./output_train.txt")

**Comment by aksnzhy** [comment]
@rybickibartek  On default, xLearn will use normalization for both training and prediction tasks. 
Can you try to train a best model by treating the test_data as validation, and then predict it on the test data, and compare the hyper-parameter tuned in CV?

**Comment by rybickibartek** [comment]
@aksnzhy I did what you told me but the result is more or less the same
#Code
import xlearn as xl

ffm_model = xl.create_ffm()              
ffm_model.setTrain("./train_ffm.txt")    
ffm_model.setValidate("./valid_ffm.txt") # here the same file
ffm_model.setTest("./valid_ffm.txt") # here the same file

param = {'task': 'reg', 'lr': 0.001, 'lambda': 0.0002, 'k': 4, 'fold': 3}

ffm_model.fit(param, "./model.out")
ffm_model.predict("./model.out", "./output_valid.txt")

#Predictions
-0.38173
-0.381636
-0.381884
-0.381659
-0.381953
-0.381829
etc.

#Program Output
----------------------------------------------------------------------------------------------
           _
          | |
     __  _| |     ___  __ _ _ __ _ __
     \ \/ / |    / _ \/ _` | '__| '_ \ 
      >  <| |___|  __/ (_| | |  | | | |
     /_/\_\_____/\___|\__,_|_|  |_| |_|

        xLearn   -- 0.40 Version --
----------------------------------------------------------------------------------------------

[------------] xLearn uses 4 threads for training task.
[ ACTION     ] Read Problem ...
[------------] First check if the text file has been already converted to binary format.
[------------] Binary file (./train_ffm.txt.bin) found. Skip converting text to binary.
[------------] First check if the text file has been already converted to binary format.
[------------] Binary file (./valid_ffm.txt.bin) found. Skip converting text to binary.
[------------] Number of Feature: 49
[------------] Number of Field: 49
[------------] Time cost for reading problem: 0.23 (sec)
[ ACTION     ] Initialize model ...
[------------] Model size: 75.42 KB
[------------] Time cost for model initial: 0.00 (sec)
[ ACTION     ] Start to train ...
[------------] Epoch      Train mse_loss       Test mse_loss     Time cost (sec)
[   10%      ]     1            7.378407            7.670690                1.13
[   20%      ]     2            7.367244            7.666442                1.11
[   30%      ]     3            7.364962            7.664942                1.12
[   40%      ]     4            7.364066            7.664250                1.12
[   50%      ]     5            7.363617            7.663892                1.11
[   60%      ]     6            7.363400            7.663667                1.12
[   70%      ]     7            7.363253            7.663536                1.12
[   80%      ]     8            7.363167            7.663437                1.12
[   90%      ]     9            7.363160            7.663378                1.12
[  100%      ]    10            7.363142            7.663335                1.12
[ ACTION     ] Start to save model ...
[------------] Model file: ./model.out
[------------] Time cost for saving model: 0.00 (sec)
[ ACTION     ] Finish training
[ ACTION     ] Clear the xLearn environment ...
[------------] Total time cost: 11.43 (sec)
----------------------------------------------------------------------------------------------
           _
          | |
     __  _| |     ___  __ _ _ __ _ __
     \ \/ / |    / _ \/ _` | '__| '_ \ 
      >  <| |___|  __/ (_| | |  | | | |
     /_/\_\_____/\___|\__,_|_|  |_| |_|

        xLearn   -- 0.40 Version --
----------------------------------------------------------------------------------------------

[------------] xLearn uses 4 threads for prediction task.
[ ACTION     ] Load model ...
[------------] Load model from ./model.out
[------------] Loss function: squared
[------------] Score function: ffm
[------------] Number of Feature: 49
[------------] Number of K: 4
[------------] Number of field: 49
[------------] Time cost for loading model: 0.00 (sec)
[ ACTION     ] Read Problem ...
[------------] First check if the text file has been already converted to binary format.
[------------] Binary file (./valid_ffm.txt.bin) found. Skip converting text to binary.
[------------] Time cost for reading problem: 0.03 (sec)
[ ACTION     ] Start to predict ...
[------------] The test loss is: 7.663341
[ ACTION     ] Clear the xLearn environment ...
[------------] Total time cost: 0.10 (sec)

**Comment by rybickibartek** [comment]
I try to normalize feature dataframe by PCA with whitening, but I still do NOT invoke disableNorm().
Now I try to make a prediction on train and valid data.

My observations below are strange. It looks that there is some problem with MSE.
I do double check. Ground truth and target in data which are used to train FFM are the same. In the case of validation data there is also no mistake.

To compute MSE I use sklearn implementation.

# Train case

command: ffm_model.predict("./model.out", "./output_train.txt")

output:
[ ACTION     ] Start to predict ...
[------------] The test loss is: 7.288925

prediction: <- it looks much more realistic then before
-0.314607
-0.157969
-0.0249284
0.0185046
-0.195276
etc.

But when I compute MSE I get 14.578019 which is around 2 * 7.288925 (declared loss in prediction). 

###

# Valid case

command: ffm_model.predict("./model.out", "./output_valid.txt")

output:
[ ACTION     ] Start to predict ...
[------------] The test loss is: 7.553453

prediction: <- it looks much more realistic then before
0.0197419
-0.0766957
-0.204431
-0.391227
-0.0799143
etc.

But when I compute MSE I get 15.106908 which is around 2 * 7.553453 (declared loss in prediction)

**Comment by rybickibartek** [comment]
I attach prefixes of both train and valid files (after PCA + whitening). When I do train and predict only on the below data all problems described above also occur.

Prefix of my train_ffm.txt file:
-0.8202825999999999 0:0:0.3117789414565015 1:1:-0.5652288517733643 2:2:-0.04554901098592678 3:3:-0.9175931866344247
0.39291325 0:0:-0.22510521313928353 1:1:0.00425873477077314 2:2:-0.4469051024042328 3:3:-0.6296660124731122
0.68805599 0:0:0.6301425801873907 1:1:0.320284793878601 2:2:0.23199666087640652 3:3:-1.5986888639868326
0.14249520000000002 0:0:0.003269064937081595 1:1:-0.5784824829882438 2:2:0.180537649203927 3:3:-0.39341058514046656
-0.15974919 0:0:0.14396043484547769 1:1:-0.9002096026116833 2:2:0.14785316573042145 3:3:0.38473376923459596
0.87158529 0:0:0.7717171880242205 1:1:1.2711225154252572 2:2:-0.7865411043210652 3:3:2.4557392262007407
0.23012899 0:0:0.05182382672077052 1:1:0.22332864244453174 2:2:-0.19245765171652668 3:3:-0.7168201563299569
2.13584976 0:0:0.22890943201070502 1:1:-0.042133261298596895 2:2:0.3250219394261894 3:3:0.40766907902855837
-0.06540639 0:0:0.224382534500724 1:1:-0.12401573877583515 2:2:0.5119601100442842 3:3:-0.6199324220204467
0.30006168 0:0:-0.029515428894727056 1:1:0.4193359885294273 2:2:0.060297167215847054 3:3:-0.9533411997027869

Prefix of my valid_ffm.txt file:
1.35976607 0:0:0.41970538915814176 1:1:-0.7146265277670586 2:2:0.453079299975555 3:3:-0.42966687645918045
0.15634038 0:0:0.45420601479094513 1:1:0.053865225854522795 2:2:0.1182247423179377 3:3:-1.4775425862965594
0.18239631 0:0:0.06591573738832948 1:1:-0.7023590724385997 2:2:0.24429872916123357 3:3:-0.23323457328324668
-0.36759344 0:0:0.04091793073952817 1:1:-0.2705448702155537 2:2:-0.2431950626278499 3:3:0.03034896388899168
1.94509685 0:0:-0.2510287615161921 1:1:-0.2786247488467375 2:2:0.12347717253770238 3:3:0.6036756371170369
0.87068986 0:0:0.12298077902023713 1:1:-0.1248925599629035 2:2:0.01707941593565715 3:3:-0.7995385696012618
-3.74021474 0:0:0.12342870417541334 1:1:-0.0843147364918249 2:2:0.18840141969956 3:3:-1.7519728981884406
-0.75123008 0:0:-0.1735614695763696 1:1:0.19113507904992097 2:2:0.02110340772887806 3:3:-1.6603270236887784
0.1507508 0:0:0.2844666892433216 1:1:0.25131040495674195 2:2:0.28731948252783346 3:3:-1.2531374532062154
1.54748932 0:0:-0.06517174429898606 1:1:-0.49880456266271456 2:2:0.27585648841711635 3:3:-0.5122646493701404

**Comment by BBBBBigRiceRoll** [comment]
I just have the same problem with you. Have you solved it now? And how you handled it?
Thank you!

**Comment by rybickibartek** [comment]
Good to know that someone has the same problem as me and I not alone ;-)
No I didn't solve it.

**Comment by AllardJM** [comment]
I get all the same predictions as well
